### PR TITLE
Add an embedded simulator to the documentation site

### PR DESCRIPTION
**Pull request by jsierles**
Appetize.io graciously offered free simulator usage both for rnplay.org and the docs. This is a preview of how embedding the simulator might work: http://quick.as/X6qlfpYj2

The goal of this PR is to get feedback on the concept and design for We'd like to make it easy for others to add more examples like this one, so eventually everything on both iOS and Android have a runnable example.

Possible changes to UIExplorer are:

* Enabling LinkingIOS deep linking (in progress, but only for iOS currently)
* Fetching from iOS UserDefaults for deep linking (working, also could work on Android)

Website changes:

* adding javascript for a modal, and the simulator frame embed
* bundling UIExplorer's javascript, to be loaded remotely by the simulator
* adding screenshots for each component

### CODE REVIEW

**Comment by jsierles** [comment]
Will look at the test failures shortly, and squash commits once we settle on a final approach.

Also, we can add support for loading on a device soon, either via the Playground app, exponent, or a dedicated UIExplorer app that loads its jsbundle from the website. If the latter is useful, we can work on building and submitting them.

@vjeux @ide

**Comment by vjeux** [comment]
So exciting! I'll take a look at it tomorrow

**Comment by vjeux** [comment]
cc @brentvatne

**Comment by christopherdro** [comment]
:+1:

File: Examples/UIExplorer/UIExplorerList.ios.js [hunk 1]
```diff
@@ -111,11 +113,33 @@ type Props = {
...
 
+  constructor(props: any) {
+    super(props);
+    this.state = {
+      route: null
+    };
+  }
+
+  componentWillMount() {
+
```
Comment (vjeux) [inline Examples/UIExplorer/UIExplorerList.ios.js]: nit remove space

File: Examples/UIExplorer/UIExplorerList.ios.js [hunk 2]
```diff
@@ -111,11 +113,33 @@ type Props = {
...
+    LinkingIOS.addEventListener('url', function() {
```
Comment (vjeux) [inline Examples/UIExplorer/UIExplorerList.ios.js]: nice :)

File: Examples/UIExplorer/UIExplorerList.ios.js [hunk 3]
```diff
@@ -111,11 +113,33 @@ type Props = {
...
+      var url = e.url.replace('uiexplorer://', '').split('?');
```
Comment (vjeux) [inline Examples/UIExplorer/UIExplorerList.ios.js]: it sucks that js doesn't come with a url parsing lib

File: Examples/UIExplorer/UIExplorerListBase.js [hunk 4]
```diff
@@ -48,7 +48,16 @@ class UIExplorerListBase extends React.Component {
...
+  componentDidUpdate() {
+    if (this.props.route !== null) {
+      var route = decodeURI(this.props.route);
+      var component = this.props.components.concat(this.props.apis).find((component) => component.title === route);
+      setTimeout(() => this.onPressRow(component), 80);
+    }
+  }
+
   render() {
+
```
Comment (vjeux) [inline Examples/UIExplorer/UIExplorerListBase.js]: remove

File: Examples/UIExplorer/UIExplorerListBase.js [hunk 5]
```diff
@@ -48,7 +48,16 @@ class UIExplorerListBase extends React.Component {
     this.search(this.state.searchText);
   }
 
+  componentDidUpdate() {
+    if (this.props.route !== null) {
+      var route = decodeURI(this.props.route);
+      var component = this.props.components.concat(this.props.apis).find((component) => component.title === route);
```
Comment (vjeux) [inline Examples/UIExplorer/UIExplorerListBase.js]: what if component doesn't exist?

File: website/layout/AutodocsLayout.js [hunk 6]
```diff
@@ -451,10 +451,53 @@ var Autodocs = React.createClass({
...
+
         </section>
       </Site>
     );
   }
 });
 
+var EmbeddedSimulator = React.createClass({
+  render: function() {
+    if ( ! this.props.shouldRender ) return null;
```
Comment (vjeux) [inline website/layout/AutodocsLayout.js]: please always use {} and put the return null; on its own line.

Also, no space inside of () and after !

File: website/layout/AutodocsLayout.js [hunk 7]
```diff
@@ -451,10 +451,53 @@ var Autodocs = React.createClass({
...
+    var encodedParams = encodeURIComponent(JSON.stringify(appParams));
+    var url = `https://appetize.io/embed/bypdk4jnjra5uwyj2kzd2aenv4?device=iphone5s&scale=70&autoplay=false&orientation=portrait&deviceColor=white&params=${encodedParams}`;
+
+    return (
+      <div>
+        <div className="modal">
+          <div className="modal-content">
+            <button className="modal-button-close">&times;</button>
+            <div className="center">
+              <iframe className="simulator" src={url} width="274px" height="550px" frameborder="0" scrolling="no"></iframe>
```
Comment (vjeux) [inline website/layout/AutodocsLayout.js]: width and height are without `px`. I'm actually surprised it worked with it :p

File: website/src/react-native/js/scripts.js [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+  modalButtonClose.addEventListener('click', hideModal);
+}
+
+function showModal(e) {
+  var backdrop = document.querySelector('.modal-backdrop');
+  var modal = document.querySelector('.modal');
+
+  backdrop.classList.add('modal-open');
+  modal.classList.add('modal-open');
+
```
Comment (vjeux) [inline website/src/react-native/js/scripts.js]: remove

File: website/src/react-native/js/scripts.js [hunk 9]
```diff
@@ -0,0 +1,29 @@
+document.addEventListener("DOMContentLoaded", init);
```
Comment (vjeux) [inline website/src/react-native/js/scripts.js]: nit: single quote

File: website/src/react-native/js/scripts.js [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+function showModal(e) {
+  var backdrop = document.querySelector('.modal-backdrop');
+  var modal = document.querySelector('.modal');
+
+  backdrop.classList.add('modal-open');
+  modal.classList.add('modal-open');
+
+}
+
+
```
Comment (vjeux) [inline website/src/react-native/js/scripts.js]: nit: only one empty line between the two functions

File: Examples/UIExplorer/UIExplorer/RNUserDefaultsIOS.m [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+@implementation RNUserDefaultsIOS
+
+RCT_EXPORT_MODULE()
+
+RCT_EXPORT_METHOD(stringForKey:(NSString *)key callback:(RCTResponseSenderBlock)callback) {
+
+    id response = [[NSUserDefaults standardUserDefaults] stringForKey:key];
+
+    if (response) {
+
```
Comment (vjeux) [inline Examples/UIExplorer/UIExplorer/RNUserDefaultsIOS.m]: rmeove empty line

File: Examples/UIExplorer/UIExplorer/RNUserDefaultsIOS.m [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+        callback(@[[NSNull null], response]);
+    }
```
Comment (vjeux) [inline Examples/UIExplorer/UIExplorer/RNUserDefaultsIOS.m]: } else {

File: Examples/UIExplorer/UIExplorer/RNUserDefaultsIOS.m [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+    else {
+
```
Comment (vjeux) [inline Examples/UIExplorer/UIExplorer/RNUserDefaultsIOS.m]: remove empty line

File: Examples/UIExplorer/UIExplorer/RNUserDefaultsIOS.m [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+//
+
+#import "RNUserDefaultsIOS.h"
+
+@implementation RNUserDefaultsIOS
+
+RCT_EXPORT_MODULE()
+
+RCT_EXPORT_METHOD(stringForKey:(NSString *)key callback:(RCTResponseSenderBlock)callback) {
+
```
Comment (vjeux) [inline Examples/UIExplorer/UIExplorer/RNUserDefaultsIOS.m]: remove empty line

File: Examples/UIExplorer/UIExplorer/AppDelegate.m [hunk 15]
```diff
@@ -61,6 +61,8 @@ - (NSURL *)sourceURLForBridge:(__unused RCTBridge *)bridge
 
     sourceURL = [NSURL URLWithString:@"http://localhost:8081/Examples/UIExplorer/UIExplorerApp.ios.bundle?platform=ios&dev=true"];
 
+    //sourceURL = [NSURL URLWithString:@"https://rnplay.org/uiexplorer.bundle"];
```
Comment (vjeux) [inline Examples/UIExplorer/UIExplorer/AppDelegate.m]: remove?

File: Examples/UIExplorer/UIExplorerList.ios.js (hunk 3 above)
Comment (jsierles) [inline Examples/UIExplorer/UIExplorerList.ios.js]: We can use this or import qs.

**Comment by vjeux** [comment]
All the comments are minor, the approach looks good.

File: Examples/UIExplorer/UIExplorerList.ios.js (hunk 3 above)
Comment (vjeux) [inline Examples/UIExplorer/UIExplorerList.ios.js]: this is fine. Please don't import a giant library just for splitting a string

**Comment by jsierles** [comment]
Added the suggested fixes and squashed commits. Tests should also pass!

File: website/src/react-native/js/scripts.js [hunk 16]
```diff
@@ -0,0 +1,31 @@
+document.addEventListener('DOMContentLoaded', init);
+
+function init() {
+  if (!document) {
```
Comment (vjeux) [inline website/src/react-native/js/scripts.js]: should be `typeof document === 'undefined'`

**Comment by vjeux** [comment]
It seems like you haven't taken all my feedback into account. Are you planing to do more changes?

**Comment by jozan** [comment]
@vjeux We'll fix those shorly. Thanks for comments!

File: Examples/UIExplorer/UIExplorer/RNUserDefaultsIOS.m (hunk 14 above)
Comment (vjeux) [inline Examples/UIExplorer/UIExplorer/RNUserDefaultsIOS.m]: ?

File: website/server/generate.js [hunk 17]
```diff
@@ -22,7 +22,7 @@ server.noconvert = true;
 // requests.
 var queue = Promise.resolve();
 
-glob('src/**/*.*', function(er, files) {
+glob('src/**/*.*', { ignore: 'src/react-native/js/**/*' }, function(er, files) {
```
Comment (vjeux) [inline website/server/generate.js]: sweet

**Comment by jsierles** [comment]
Closing in favor of #3419 and #3420.

**Pull request closed without merging**